### PR TITLE
Force stop iteration after local response is sent

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: mathetake <takeshi@tetrate.io>

ref
- https://github.com/envoyproxy/envoy/pull/13930
- https://github.com/envoyproxy/envoy/issues/13857

### CODE REVIEW

**Comment by mathetake** [comment]
I made the change almost identical to https://github.com/envoyproxy/envoy/pull/13873 but something weird seems to happen here:

with the change in Envoy, the decorder/encoder message like this and it looks fine 
```
[2020-11-03 16:21:04.041][871494][trace][http] [source/common/http/filter_manager.cc:938] [C0][S552029742600055844] encode headers called: filter=0x3f33ffcd47d0 status=0
[2020-11-03 16:21:04.041][871494][trace][http] [source/common/http/filter_manager.cc:1080] [C0][S552029742600055844] encode data called: filter=0x3f33ffcd47d0 status=0
[2020-11-03 16:21:04.041][871494][trace][http] [source/common/http/filter_manager.cc:451] [C0][S552029742600055844] decode headers called: filter=0x3f33fea12540 status=1
```

but with the change here, we have the following logs, and the response never is never returned to the client (curl in this case)

```
envoy --concurrency 1 -l trace -c examples/http_auth_random/envoy.yaml 2>&1 | grep "called:"
[2020-11-03 16:26:52.872][887837][trace][http] [source/common/http/filter_manager.cc:938] [C0][S6785004984496335233] encode headers called: filter=0xe013e0a8370 status=1
[2020-11-03 16:26:52.872][887837][trace][http] [source/common/http/filter_manager.cc:1080] [C0][S6785004984496335233] encode data called: filter=0xe013e0a8370 status=3
[2020-11-03 16:26:52.872][887837][trace][http] [source/common/http/filter_manager.cc:451] [C0][S6785004984496335233] decode headers called: filter=0xe013ea07f20 status=1
```

File: src/exports.cc [hunk 1]
```diff
@@ -186,6 +186,7 @@ Word send_local_response(void *raw_context, Word response_code, Word response_co
   auto additional_headers = toPairs(additional_response_header_pairs.value());
   context->sendLocalResponse(response_code, body.value(), std::move(additional_headers), grpc_code,
                              details.value());
+  context->wasm()->addAfterVmCallAction([context] { context->stopIteration(); });
```
Comment (PiotrSikora) [inline src/exports.cc:189]: Maybe let's call `stopIteration()` from within Envoy's `Context::sendLocalResponse()`? This seems a bit weird here.

File: src/context.cc [hunk 2]
```diff
@@ -469,44 +469,62 @@ template <typename P> static uint32_t headerSize(const P &p) { return p ? p->siz
...
+  uint64_t result;
+  {
+    DeferAfterCallActions actions(this);
+    result = wasm_->on_request_headers_abi_01_
+                 ? wasm_->on_request_headers_abi_01_(this, id_, headers).u64_
+                 : wasm_
+                       ->on_request_headers_abi_02_(this, id_, headers,
+                                                    static_cast<uint32_t>(end_of_stream))
+                       .u64_;
+  }
```
Comment (PiotrSikora) [inline src/context.cc:481]: Adding scope here will fire off all the deferred calls in the middle of this function. This significantly changes the flow control, and it's going to make bugs hard to reason about. I think you should revert those changes.

**Comment by mathetake** [comment]
the above issue is resolved by delegating the call of stopIteration into Envoy side: https://github.com/envoyproxy/envoy/blob/master/source/extensions/common/wasm/context.cc#L1571

**Comment by mathetake** [comment]
same problem lingers on.......

**Comment by mathetake** [comment]
it seems to me that OnResponseHeaders and OnResponseBody are called in response to the sendLocalResponse?

**Comment by mathetake** [comment]
I'm know next to nothing about the internal behavior of local response, but the logs show that the filter addresses(?) are different

**Comment by PiotrSikora** [comment]
> it seems to me that OnResponseHeaders and OnResponseBody are called in response to the sendLocalResponse? And the BaseContext used to handle these events must be the same the context which invoked sendLocalResponse, which means `stop_iteration_` is true for the local response as well. I think that is the difference and something unexpected seems happening

Yes, `OnResponse{Headers,Body,Trailers}` are called for the local response as well.

File: src/context.cc [hunk 3]
```diff
@@ -636,6 +614,40 @@ WasmResult ContextBase::setTimerPeriod(std::chrono::milliseconds period,
   return WasmResult::Ok;
 }
 
+FilterHeadersStatus ContextBase::convertVmCallResultToFilterHeadersStatus(uint64_t result) {
+  if (stop_iteration_) {
+    stop_iteration_ = false;
+    return FilterHeadersStatus::StopIteration;
+  } else if (result > static_cast<uint64_t>(FilterHeadersStatus::StopAllIterationAndWatermark)) {
+    return FilterHeadersStatus::StopAllIterationAndWatermark;
```
Comment (PiotrSikora) [inline src/context.cc:622]: Any reason for not pulling `stop_iteration_` case here? Does it break if we return `StopAllIterationAndWatermark`?

File: src/context.cc [hunk 4]
```diff
@@ -636,6 +614,40 @@ WasmResult ContextBase::setTimerPeriod(std::chrono::milliseconds period,
...
+    return FilterDataStatus::StopIterationNoBuffer;
+  }
+  return static_cast<FilterDataStatus>(result);
+}
+
+FilterTrailersStatus ContextBase::convertVmCallResultToFilterTrailersStatus(uint64_t result) {
+  if (!stop_iteration_ &&
+      static_cast<FilterTrailersStatus>(result) == FilterTrailersStatus::Continue) {
+    return FilterTrailersStatus::Continue;
+  }
```
Comment (PiotrSikora) [inline src/context.cc:639]: This is a bit inconsistent with other functions, could you change it to match the style? i.e.
```
if (stop_iteration_ || result > static_cast<uint64_t>(FilterTrailersStatus::StopIteration) {
    stop_iteration_ = false;
    return FilterTrailersStatus::StopIteration;
}
return static_cast<FilterTrailersStatus>(result);
```

File: src/context.cc (hunk 3 above)
Comment (mathetake) [reply src/context.cc:621]: I just followed the implementation of router https://github.com/envoyproxy/envoy/blob/master/source/common/router/router.cc#L367-L369 but I verified it's working even if return `StopAllIterationAndWatermark`.

File: include/proxy-wasm/context.h [hunk 5]
```diff
@@ -170,6 +170,11 @@ class ContextBase : public RootInterface,
   // Called before deleting the context.
   virtual void destroy();
 
+  // Called to raise the flag which indicates that the context should stop iteration regardless of
+  // returned filter status from WASM extensions. For example, we ignore
```
Comment (PiotrSikora) [inline include/proxy-wasm/context.h:174]: `s/WASM/Proxy-Wasm/`

**Review by PiotrSikora — COMMENTED** [review]
This looks great, thanks! Although, I wonder if we should be so permissive and "fix" the broken code on the fly.

Similarly, could you open a PR in Envoy against your branch to verify that it passes all tests?

**Comment by mathetake** [comment]
>  Although, I wonder if we should be so permissive and "fix" the broken code on the fly.

Agreed

**Comment by mathetake** [comment]
@PiotrSikora  https://github.com/envoyproxy/envoy/pull/13930 all tests passed!

**Review by PiotrSikora — APPROVED** [review]

**Comment by PiotrSikora** [comment]
@mathetake It looks that we don't have good coverage, since this breaks send local reply in production code.

**Comment by PiotrSikora** [comment]
@mathetake OK, I see the issue now. We're overriding "current" action with `StopAllIterationAndWatermark`, however, not all callbacks return action (most notably, `on_http_call_response` does not), in which case we're not overriding "current" action, but action of the subsequent call. We need to clear `stop_iteration_` flag after we return from VM, not when converting actions.

**Pull request merged**